File: shared/diagram-renderer/src/views/behavior-common.ts
import type { Selection } from "d3";
import ELK from "elkjs/lib/elk.bundled.js";
import type { PreparedNode, PreparedView } from "../prepare";
import type { RenderOptions } from "../renderer";
import type { DiagramTheme } from "../theme";
import { collectElkEdgeLabels, edgeLabelPositionFromSections, estimateElkLabelBox, type ElkLabelBox } from "./elk-label-utils";

export { edgeLabelPositionFromSections };

export interface EdgeSection {
  startPoint?: { x: number; y: number };
  bendPoints?: Array<{ x: number; y: number }>;
  endPoint?: { x: number; y: number };
}

export const behaviorElk = new ELK();

export interface LaidOutRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface BehaviorLayoutResult {
  positions: Map<string, LaidOutRect>;
  edgeSectionsById: Map<string, EdgeSection[]>;
  edgeLabelsById: Map<string, ElkLabelBox[]>;
}

export interface BehaviorSceneContext {
  root: Selection<SVGGElement, unknown, null, undefined>;
  prepared: PreparedView;
  theme: DiagramTheme;
  width: number;
  height: number;
  options?: RenderOptions;
}

export function nodeKind(node: PreparedNode): string {
  return String(node.kind || "action").toLowerCase();
}

export function pathFromSections(sections: EdgeSection[] | undefined): string | null {
  if (!sections?.length) return null;
  const parts: string[] = [];
  for (const section of sections) {
    if (!section.startPoint || !section.endPoint) continue;
    parts.push(`M${section.startPoint.x},${section.startPoint.y}`);
    for (const point of section.bendPoints ?? []) {
      parts.push(`L${point.x},${point.y}`);
    }
    parts.push(`L${section.endPoint.x},${section.endPoint.y}`);
  }
  return parts.length ? parts.join(" ") : null;
}

export function buildSelfLoopPath(node: LaidOutRect): { path: string; labelX: number; labelY: number } {
  const startX = node.x + node.width;
  const startY = node.y + node.height / 2 - 8;
  const loopRadius = 28;
  return {
    path:
      `M${startX},${startY}` +
      ` C${startX + loopRadius},${startY - loopRadius}` +
      ` ${startX + loopRadius},${startY + loopRadius}` +
      ` ${startX},${startY + 18}`,
    labelX: startX + loopRadius + 8,
    labelY: startY,
  };
}

export function fallbackEdgePath(
  source: LaidOutRect,
  target: LaidOutRect,
  horizontal: boolean,
): { path: string; labelX: number; labelY: number } {
  if (
    source.x === target.x &&
    source.y === target.y &&
    source.width === target.width &&
    source.height === target.height
  ) {
    return buildSelfLoopPath(source);
  }
  if (horizontal) {
    const startX = source.x + source.width;
    const startY = source.y + source.height / 2;
    const endX = target.x;
    const endY = target.y + target.height / 2;
    const midX = (startX + endX) / 2;
    return {
      path: `M${startX},${startY} L${midX},${startY} L${midX},${endY} L${endX},${endY}`,
      labelX: midX,
      labelY: (startY + endY) / 2 - 6,
    };
  }
  const startX = source.x + source.width / 2;
  const startY = source.y + source.height;
  const endX = target.x + target.width / 2;
  const endY = target.y;
  const midY = (startY + endY) / 2;
  return {
    path: `M${startX},${startY} L${startX},${midY} L${endX},${midY} L${endX},${endY}`,
    labelX: (startX + endX) / 2,
    labelY: midY - 6,
  };
}

function nodeDimensions(node: PreparedNode, mode: "action" | "state"): { width: number; height: number } {
  const kind = nodeKind(node);
  if (mode === "state") {
    if (kind.includes("initial") || kind.includes("final")) return { width: 34, height: 34 };
    if (kind.includes("composite")) return { width: 340, height: 320 };
    return { width: 240, height: 180 };
  }
  if (kind.includes("initial") || kind.includes("final") || kind.includes("start") || kind.includes("done")) {
    return { width: 40, height: 40 };
  }
  if (kind.includes("decision") || kind.includes("merge")) return { width: 76, height: 76 };
  if (kind.includes("fork") || kind.includes("join")) return { width: 220, height: 14 };
  return { width: 220, height: 68 };
}

function transitionDisplayLabel(label: string): string {
  const trimmed = label.trim();
  if (!trimmed || trimmed.toLowerCase() === "entry") return "";
  return trimmed;
}

export async function layoutBehaviorGraph(
  prepared: PreparedView,
  options: { horizontal?: boolean; mode: "action" | "state" },
): Promise<BehaviorLayoutResult> {
  const horizontal = options.horizontal ?? false;
  const positions = new Map<string, LaidOutRect>();
  const edgeSectionsById = new Map<string, EdgeSection[]>();
  const edgeLabelsById = new Map<string, ElkLabelBox[]>();

  const children = prepared.nodes.map((node) => {
    const size = nodeDimensions(node, options.mode);
    return { id: node.id, width: size.width, height: size.height };
  });

  const edges = prepared.edges.map((edge) => {
    const displayLabel = transitionDisplayLabel(edge.label);
    const base = {
      id: edge.id,
      sources: [edge.source],
      targets: [edge.target],
    };
    if (!displayLabel) {
      return base;
    }
    const labelBox = estimateElkLabelBox(`${edge.id}::label`, displayLabel, {
      minWidth: 38,
      minHeight: 16,
      paddingX: 8,
      paddingY: 6,
      charWidth: 6,
    });
    return {
      ...base,
      labels: [
        {
          id: labelBox.id,
          text: labelBox.text,
          width: labelBox.width,
          height: labelBox.height,
          layoutOptions: {
            "org.eclipse.elk.edgeLabels.placement": "CENTER",
            "org.eclipse.elk.edgeLabels.inline": "false",
          },
        },
      ],
    };
  });

  const isState = options.mode === "state";
  const graph = {
    id: prepared.title || "behavior",
    layoutOptions: isState
      ? {
          "elk.algorithm": "layered",
          "elk.direction": horizontal ? "RIGHT" : "DOWN",
          "elk.hierarchyHandling": "INCLUDE_CHILDREN",
          "elk.edgeRouting": "ORTHOGONAL",
          "elk.layered.nodePlacement.strategy": "NETWORK_SIMPLEX",
          "elk.layered.crossingMinimization.strategy": "LAYER_SWEEP",
          "elk.layered.spacing.nodeNodeBetweenLayers": "230",
          "elk.spacing.nodeNode": "190",
          "elk.spacing.edgeNode": "130",
          "elk.spacing.edgeEdge": "110",
          "elk.spacing.edgeLabel": "12",
          "elk.padding": "[top=100,left=90,bottom=90,right=90]",
          "elk.separateConnectedComponents": "true",
          "elk.json.edgeCoords": "ROOT",
        }
      : {
          "elk.algorithm": "layered",
          "elk.direction": horizontal ? "RIGHT" : "DOWN",
          "elk.edgeRouting": "ORTHOGONAL",
          "elk.layered.nodePlacement.strategy": "NETWORK_SIMPLEX",
          "elk.layered.crossingMinimization.strategy": "LAYER_SWEEP",
          "elk.spacing.nodeNode": horizontal ? "90" : "120",
          "elk.layered.spacing.nodeNodeBetweenLayers": horizontal ? "190" : "170",
          "elk.spacing.edgeNode": "80",
          "elk.spacing.edgeEdge": "60",
          "elk.spacing.edgeLabel": "12",
          "elk.padding": "[top=80,left=80,bottom=80,right=80]",
          "elk.separateConnectedComponents": "true",
          "elk.json.edgeCoords": "ROOT",
        },
    children,
    edges,
  };

  const laidOut = await behaviorElk.layout(graph);
  for (const child of laidOut.children ?? []) {
    positions.set(String(child.id), {
      x: child.x ?? 0,
      y: child.y ?? 0,
      width: child.width ?? 200,
      height: child.height ?? 80,
    });
  }
  for (const edge of laidOut.edges ?? []) {
    if (edge.sections) {
      edgeSectionsById.set(String(edge.id), edge.sections as EdgeSection[]);
    }
  }
  collectElkEdgeLabels(laidOut, { x: 0, y: 0 }, edgeLabelsById);

  return { positions, edgeSectionsById, edgeLabelsById };
}

export function truncateLabel(text: string, max: number): string {
  const trimmed = text.trim();
  return trimmed.length > max ? `${trimmed.slice(0, max - 2)}..` : trimmed;
}

File: crates/diagram_draw/src/behavior_common.rs
//! Port of the drawing-relevant half of `views/behavior-common.ts`, shared by `action_flow.rs`,
//! `state_transition.rs`, and (for `PreparedView`/`PreparedNode`/`PreparedEdge` only) `sequence.rs`.
//! Layout (`buildBehaviorElkGraphInput`/`layoutBehaviorGraph`) lives in the crate's private
//! `layout` module.

use std::collections::BTreeMap;

use serde::Deserialize;
use serde_json::Value;

use crate::svg::format_number as n;
use crate::types::{EdgeSection, Point};

#[derive(Debug, Deserialize)]
pub struct PreparedNode {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub kind: String,
    #[serde(default)]
    pub attributes: BTreeMap<String, Value>,
}

#[derive(Debug, Deserialize)]
pub struct PreparedEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    pub label: String,
    #[serde(rename = "edgeKind", default)]
    pub edge_kind: Option<String>,
    #[serde(default)]
    pub attributes: BTreeMap<String, Value>,
}

#[derive(Debug, Deserialize)]
pub struct PreparedView {
    pub title: String,
    pub view: String,
    pub nodes: Vec<PreparedNode>,
    pub edges: Vec<PreparedEdge>,
    #[serde(default)]
    pub meta: Value,
}

#[derive(Debug, Clone, Copy, Deserialize)]
pub struct LaidOutRect {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Deserialize)]
pub struct ElkLabelBox {
    #[allow(dead_code)]
    pub id: String,
    #[allow(dead_code)]
    pub text: String,
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Deserialize)]
pub struct BehaviorLayoutResult {
    pub positions: BTreeMap<String, LaidOutRect>,
    #[serde(rename = "edgeSectionsById", default)]
    pub edge_sections_by_id: BTreeMap<String, Vec<EdgeSection>>,
    #[serde(rename = "edgeLabelsById", default)]
    pub edge_labels_by_id: BTreeMap<String, Vec<ElkLabelBox>>,
}

/// Port of `nodeKind`: `String(node.kind || "action").toLowerCase()`.
pub fn node_kind(node: &PreparedNode) -> String {
    if node.kind.is_empty() {
        "action".to_string()
    } else {
        node.kind.to_lowercase()
    }
}

fn section_points(section: &EdgeSection) -> Vec<Point> {
    let mut points = Vec::new();
    if let Some(start) = section.start_point {
        points.push(start);
    }
    points.extend(section.bend_points.iter().copied());
    if let Some(end) = section.end_point {
        points.push(end);
    }
    points
}

/// Port of `pathFromSections`.
pub fn path_from_sections(sections: Option<&[EdgeSection]>) -> Option<String> {
    let sections = sections?;
    if sections.is_empty() {
        return None;
    }
    let mut parts = Vec::new();
    for section in sections {
        let (Some(start), Some(end)) = (section.start_point, section.end_point) else {
            continue;
        };
        parts.push(format!("M{},{}", n(start.x), n(start.y)));
        for point in &section.bend_points {
            parts.push(format!("L{},{}", n(point.x), n(point.y)));
        }
        parts.push(format!("L{},{}", n(end.x), n(end.y)));
    }
    if parts.is_empty() {
        None
    } else {
        Some(parts.join(" "))
    }
}

pub struct FallbackPath {
    pub path: String,
    pub label_x: f64,
    pub label_y: f64,
}

/// Port of `buildSelfLoopPath`.
pub fn build_self_loop_path(node: &LaidOutRect) -> FallbackPath {
    let start_x = node.x + node.width;
    let start_y = node.y + node.height / 2.0 - 8.0;
    let loop_radius = 28.0;
    FallbackPath {
        path: format!(
            "M{},{} C{},{} {},{} {},{}",
            n(start_x),
            n(start_y),
            n(start_x + loop_radius),
            n(start_y - loop_radius),
            n(start_x + loop_radius),
            n(start_y + loop_radius),
            n(start_x),
            n(start_y + 18.0)
        ),
        label_x: start_x + loop_radius + 8.0,
        label_y: start_y,
    }
}

/// Port of `fallbackEdgePath`.
pub fn fallback_edge_path(
    source: &LaidOutRect,
    target: &LaidOutRect,
    horizontal: bool,
) -> FallbackPath {
    if source.x == target.x
        && source.y == target.y
        && source.width == target.width
        && source.height == target.height
    {
        return build_self_loop_path(source);
    }
    if horizontal {
        let start_x = source.x + source.width;
        let start_y = source.y + source.height / 2.0;
        let end_x = target.x;
        let end_y = target.y + target.height / 2.0;
        let mid_x = (start_x + end_x) / 2.0;
        return FallbackPath {
            path: format!(
                "M{},{} L{},{} L{},{} L{},{}",
                n(start_x),
                n(start_y),
                n(mid_x),
                n(start_y),
                n(mid_x),
                n(end_y),
                n(end_x),
                n(end_y)
            ),
            label_x: mid_x,
            label_y: (start_y + end_y) / 2.0 - 6.0,
        };
    }
    let start_x = source.x + source.width / 2.0;
    let start_y = source.y + source.height;
    let end_x = target.x + target.width / 2.0;
    let end_y = target.y;
    let mid_y = (start_y + end_y) / 2.0;
    FallbackPath {
        path: format!(
            "M{},{} L{},{} L{},{} L{},{}",
            n(start_x),
            n(start_y),
            n(start_x),
            n(mid_y),
            n(end_x),
            n(mid_y),
            n(end_x),
            n(end_y)
        ),
        label_x: (start_x + end_x) / 2.0,
        label_y: mid_y - 6.0,
    }
}

/// Port of `truncateLabel`. JS `String.length`/`slice` are UTF-16 code units, not Unicode scalar
/// values -- match with `encode_utf16` rather than `chars()` (unlike General View's
/// `truncate_to_chars`, which never needed non-BMP parity).
pub fn truncate_label(text: &str, max: usize) -> String {
    let trimmed = text.trim();
    let units: Vec<u16> = trimmed.encode_utf16().collect();
    if units.len() > max {
        let truncated = String::from_utf16_lossy(&units[..max.saturating_sub(2)]);
        format!("{truncated}..")
    } else {
        trimmed.to_string()
    }
}

/// Port of `edgeLabelPositionFromSections`.
pub fn edge_label_position_from_sections(sections: Option<&[EdgeSection]>) -> Option<(f64, f64)> {
    let section = sections?.first()?;
    if section.start_point.is_none() || section.end_point.is_none() {
        return None;
    }
    let points = section_points(section);
    let mid_index = (points.len() - 1) / 2;
    let start = points[mid_index];
    let end = points.get(mid_index + 1).copied().unwrap_or(start);
    Some(((start.x + end.x) / 2.0, (start.y + end.y) / 2.0 - 6.0))
}
